## Import Necessary Libraries
* **numpy** : `numpy` is a fundamental Python library for numerical computing, providing support for large, multi-dimensional arrays and matrices, along with a wide range of mathematical functions to operate on them.
* **pandas** : `pandas` is a powerful Python library for data manipulation and analysis, providing data structures like DataFrames and Series to efficiently handle and analyze structured data.
* **re** : The `re` module in Python provides support for regular expressions, allowing for pattern matching, searching, and manipulating strings.


In [2]:
import numpy as np
import pandas as pd
import re

## Importing the scraped data (*mobiles_data*) which is stored in csv file

In [2]:
data = pd.read_csv(r"C:\Users\DELL\Desktop\Shashank_DataScience_Project\mobiles_data.csv")

## Basic inspection of data
* The basic inspection is done so that we can get the information about tha data like its shape, columns, data types, etc.

### Display the first few rows of the DataFrame using the head() method

In [3]:
data.head()

Name Specs Score    Rating   Price   RAM RAM Type  \
0             realme GT 6T         88%  8.0 / 10  30,999  8 GB  LPDDR5X   
1     OnePlus Nord CE 4 5G         86%   4.3 / 5  24,998  8 GB  LPDDR4X   
2    Samsung Galaxy F55 5G         89%   3.4 / 5  26,999  8 GB      NaN   
3     Infinix GT 20 Pro 5G         89%   4.4 / 5  24,999  8 GB  LPDDR5X   
4  Motorola Edge 50 Pro 5G         90%   4.4 / 5  31,142  8 GB      NaN   

            Resolution Refresh Rate       Display Type  \
0  1264x2780 px (FHD+)       120 Hz        LTPO AMOLED   
1  1080x2412 px (FHD+)       120 Hz             AMOLED   
2  1080x2400 px (FHD+)       120 Hz  Super AMOLED Plus   
3  1080x2436 px (FHD+)       144 Hz             AMOLED   
4  1220x2712 px (FHD+)       144 Hz             P-OLED   

              Screen Size  ...       Internal Memory Expandable Memory  \
0  6.78 inches (17.22 cm)  ...  128 GBBest in Class▾                No   
1   6.7 inches (17.02 cm)  ...  128 GBBest in Class▾   Yes, Up to 1 TB   
2   6.7 inches (17.02 cm)  ...  128 GBBest in Class▾   Yes, Up to 1 TB   
3  6.78 inches (17.22 cm)  ...  256 GBBest in Class▾                No   
4   6.7 inches (17.02 cm)  ...  256 GBBest in Class▾                No   

            SIM Slot                                               WiFi  \
0  Dual SIM, GSM+GSM     Yes, Wi-Fi 6E (802.11 a/b/g/n/ac/ax) 5GHz 6GHz   
1  Dual SIM, GSM+GSM  Yes, Wi-Fi 6E (802.11 a/b/g/n/ac/ax) 5GHz 6GHz...   
2  Dual SIM, GSM+GSM  Yes, Wi-Fi 6E (802.11 a/b/g/n/ac/ax) 5GHz 6GHz...   
3  Dual SIM, GSM+GSM     Yes, Wi-Fi 6E (802.11 a/b/g/n/ac/ax) 5GHz 6GHz   
4  Dual SIM, GSM+GSM     Yes, Wi-Fi 6E (802.11 a/b/g/n/ac/ax) 5GHz 6GHz   

                  WiFi Features  Bluetooth                      GPS  NFC  \
0                Mobile Hotspot  Yes, v5.4  Yes with A-GPS, Glonass  Yes   
1                Mobile Hotspot  Yes, v5.4  Yes with A-GPS, Glonass  NaN   
2  Wi-Fi Direct, Mobile Hotspot  Yes, v5.2         Yes with Glonass  Yes   
3                Mobile Hotspot        Yes           Yes with A-GPS  Yes   
4                Mobile Hotspot  Yes, v5.4  Yes with A-GPS, Glonass  Yes   

  Radio Fingerprint Sensor  
0    No                Yes  
1    No                Yes  
2    No                Yes  
3   NaN                Yes  
4    No                Yes  

[5 rows x 28 columns]

### Display the last few rows of the DataFrame using the head() method

In [4]:
data.tail()

Name Specs Score   Rating                     Price  \
13727   VOX Mobile VGS-603         45%  4.1 / 5   1,609(Last Known Price)   
13728           Onida F950         86%      NaN   1,099(Last Known Price)   
13729           HTC Tattoo         45%  1.9 / 5  22,900(Last Known Price)   
13730          Karbonn K20         45%   05-May   1,489(Last Known Price)   
13731  Samsung S5620 Monte         45%  4.3 / 5  10,200(Last Known Price)   

          RAM RAM Type        Resolution Refresh Rate Display Type  \
13727     NaN      NaN  240 x 400 pixels          NaN          LCD   
13728     NaN      NaN  176 x 220 pixels          NaN          LCD   
13729  256 MB      NaN  240 x 320 pixels          NaN          LCD   
13730     NaN      NaN  320 x 240 pixels          NaN          NaN   
13731     NaN      NaN  240 x 400 pixels          NaN          LCD   

                Screen Size  ... Internal Memory Expandable Memory  \
13727    3 inches (7.62 cm)  ...             NaN   Yes, Up to 2 GB   
13728  1.9 inches (4.83 cm)  ...             NaN   Yes, Up to 2 GB   
13729  2.8 inches (7.11 cm)  ...          512 MB               Yes   
13730   2.4 inches (6.1 cm)  ...          259 KB   Yes, Up to 8 GB   
13731    3 inches (7.62 cm)  ...          227 MB  Yes, Up to 16 GB   

                SIM Slot                         WiFi WiFi Features  \
13727  Dual SIM, GSM+GSM                           No           NaN   
13728    Single SIM, GSM                           No           NaN   
13729    Single SIM, GSM  Yes, Wi-Fi 3 (802.11 a/b/g)           NaN   
13730  Dual SIM, GSM+GSM                           No           NaN   
13731    Single SIM, GSM    Yes, Wi-Fi 3 (802.11 b/g)           NaN   

       Bluetooth             GPS  NFC     Radio Fingerprint Sensor  
13727    Yes, v2             NaN  NaN       Yes                NaN  
13728        Yes             NaN  NaN       NaN                NaN  
13729    Yes, v2  Yes with A-GPS  NaN  Yes, RDS                NaN  
13730        Yes             NaN  NaN       Yes                NaN  
13731  Yes, v2.1             NaN  NaN  Yes, RDS                NaN  

[5 rows x 28 columns]

### Checking summary of a DataFrame using .info() method

In [182]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12904 entries, 0 to 13731
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Name                12901 non-null  object
 1   Specs Score         12902 non-null  object
 2   Rating              7047 non-null   object
 3   Price               12902 non-null  object
 4   RAM                 9346 non-null   object
 5   RAM Type            3017 non-null   object
 6   Resolution          12895 non-null  object
 7   Refresh Rate        2032 non-null   object
 8   Display Type        12618 non-null  object
 9   Screen Size         12775 non-null  object
 10  Chipset             7344 non-null   object
 11  CPU                 8950 non-null   object
 12  Architecture        5089 non-null   object
 13  Fabrication         3168 non-null   object
 14  Graphics            7028 non-null   object
 15  Pixel Density       12590 non-null  object
 16  Quick Charging      32

### Get the shape DataFrame using the .shape attribute

In [177]:
data.shape

(13732, 28)

In [11]:
# Number of rows in Dataset
rows = data.shape[0]
print(f"The imported/uncleaned mobiles data have {rows} no. of rows")

The imported/uncleaned mobiles data have 13732 no. of rows


In [12]:
# Number of columns in Dataset
cloumns = data.shape[1]
print(f"The imported/uncleaned mobiles data have {cloumns} no. of columns")

The imported/uncleaned mobiles data have 28 no. of columns


In [178]:
#remove duplicates
#handle missing values
#dealing with outliers (not for data analysis)
#converting data types
#handling categorical data
#dealing with insonsistence data

## Handling duplicated values

In [179]:
#Checking if there are an duplicated values
data.duplicated().sum()

828

In [180]:
# Dropping the tuples/rows which are duplicated
data.drop_duplicates(inplace=True)

## Handling missing values

In [183]:
# Checking if there are any missing values
data.isna().sum()

Name                      3
Specs Score               2
Rating                 5857
Price                     2
RAM                    3558
RAM Type               9887
Resolution                9
Refresh Rate          10872
Display Type            286
Screen Size             129
Chipset                5560
CPU                    3954
Architecture           7815
Fabrication            9736
Graphics               5876
Pixel Density           314
Quick Charging         9617
Type-C                 6552
Internal Memory        2996
Expandable Memory       107
SIM Slot                 16
WiFi                   2855
WiFi Features          5479
Bluetooth                71
GPS                    4519
NFC                    7792
Radio                  3092
Fingerprint Sensor     6362
dtype: int64

* After checking the missing values we get to know that:
    * We have 11 columns having more than 5000 missing values
    * Some of those columns are not really necessary for the further work
    * But, the rating column is important and also have 5857 null values
    * Thus, we will not drop the 'rating' column
* Thus before handling missing values, we will delete the columns having missing values more than 5000 except 'rating' column.

In [185]:
# Deleting the columns having more than 5000 missing values
data.drop(["Refresh Rate","RAM Type","Fabrication","Architecture","Quick Charging","Type-C","NFC","Fingerprint Sensor","WiFi Features","Graphics"],axis=1,inplace=True)

#### Checking the missing values after the deletion of columns

In [189]:
temp = data
temp.isna().sum()

Name                    3
Specs Score             2
Rating               5857
Price                   2
RAM                  3558
Resolution              9
Display Type          286
Screen Size           129
Chipset              5560
CPU                  3954
Pixel Density         314
Internal Memory      2996
Expandable Memory     107
SIM Slot               16
WiFi                 2855
Bluetooth              71
GPS                  4519
Radio                3092
dtype: int64

#### Dropping the null values from the dataset 
* Here we will not drop null values from 'rating' column
* We will further replace the null values in 'rating' column with 0.

In [190]:
tempdata = temp.dropna(subset=["CPU","Name","Specs Score","Price","RAM","Resolution","Display Type","Screen Size","Chipset","Pixel Density","Internal Memory","Expandable Memory","SIM Slot","WiFi","Bluetooth","GPS","Radio"])

In [192]:
tempdata.isna().sum()

Name                    0
Specs Score             0
Rating               1200
Price                   0
RAM                     0
Resolution              0
Display Type            0
Screen Size             0
Chipset                 0
CPU                     0
Pixel Density           0
Internal Memory         0
Expandable Memory       0
SIM Slot                0
WiFi                    0
Bluetooth               0
GPS                     0
Radio                   0
dtype: int64

## Replacing the null values in 'Rating' column with zero(0)

In [193]:
tempdata.loc[(tempdata['Rating'].isna(),'Rating')] = 0

C:\Users\DELL\AppData\Local\Temp\ipykernel_9632\977108005.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tempdata.loc[(tempdata['Rating'].isna(),'Rating')] = 0


## Cleaning the columns from the dataset

In [196]:
df = tempdata

### Cleaning the 'Specs Score' column

#### Remove '%' sign from specs score column and change the column datatype to 'float'

In [198]:
df["Specs Score"] = df["Specs Score"].str.replace("%","")
df["Specs Score"] = df["Specs Score"].astype(float)

C:\Users\DELL\AppData\Local\Temp\ipykernel_9632\1018227404.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Specs Score"] = df["Specs Score"].str.replace("%","")
C:\Users\DELL\AppData\Local\Temp\ipykernel_9632\1018227404.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Specs Score"] = df["Specs Score"].astype(float)


## Cleaning the 'Rating' Column

#### Spliting the rating column from '/' and only take first value
* We have Rating column in format of (8/10,2/3)
* Splitting the 'Rating column and teking the value before the ('/')

In [200]:
df['Rating'] = df['Rating'].str.split('/').str[0]
df.loc[(df['Rating'].isna(),'Rating')] = 0

C:\Users\DELL\AppData\Local\Temp\ipykernel_9632\2969385273.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Rating'] = df['Rating'].str.split('/').str[0]


#### Converting the data-type to 'folat32'

In [203]:
df['Rating'] = df['Rating'].astype('float32')

C:\Users\DELL\AppData\Local\Temp\ipykernel_9632\3196477549.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Rating'] = df['Rating'].astype('float32')


#### Converting the ratings which are out of 10 to out of 5
* We have most of the rating out of 5 but, some of those are out of 10
* So we are dividing the ratings which are greater than 5 by 2 so that the rating which are out of 10 will becomes out of 5

In [205]:
df.loc[(df['Rating'] > 5,'Rating')] = (df.loc[(df['Rating'] > 5,'Rating')])/2

### Cleaning the 'Price' column

In [ ]:
pattern = re.compile(r"[1234567890]")
x = df[df['Price'].str.contains("Notify")]["Price"].apply(lambda x: pattern.findall(x)).str.join("")

In [ ]:
df[df['Price'].str.contains("Notify")]["Price"] = x

In [ ]:
pattern = re.compile(r"[1234567890]")
mask = df['Price'].str.contains("Notify",na=False)
df.loc[mask, 'Price'] = df.loc[mask, 'Price'].apply(lambda x: pattern.findall(x)).str.join("")

In [ ]:
def temp(x):
    try:
        if int(x):
            return False
    except:
        return True

In [ ]:
mask = df['Price'].apply(temp)
df.loc[mask,"Price"]=  df.loc[mask,"Price"].str.join("")

In [ ]:
df['Price']=df['Price'].astype("int")

In [ ]:
df.drop(index = df[df['Price'] > 800000].index,inplace=True)

In [215]:
df.to_csv('temp1_mobile_data.csv',index=False)

------------------

In [216]:
data = pd.read_csv(r"C:\Users\DELL\Desktop\Shashank_DataScience_Project\temp1_mobile_data.csv")

### Creating a new 'brand_name' column
* As per the observation 'Name' column containing the name of the mobiles have brand name at very first word
* Extracting the brand name from the name of the mobiles

In [165]:
df['brand_name'] = df["Name"].str.split(' ').str[0]

### Cleaning the 'RAM' column
* There are some phones whose ram is in MB and there are some whose ram is in GB, so lets convert all of them in same format

In [ ]:
df.loc[df["RAM"].str.split(" ").str.get(1) == 'GB',"RAM"] = df[df["RAM"].str.split(" ").str.get(1) == 'GB']['RAM'].str.split(" ").str.get(0).apply(lambda x: float(x) * 1000)

In [ ]:
df.loc[df["RAM"].str.contains("MB",na=False),"RAM"] = df.loc[df["RAM"].str.contains("MB",na=False),"RAM"].str.split(" ").str.get(0)

In [ ]:
df.loc[(df['RAM'].str.contains("RAM",na=False)),"RAM"] = df[df['RAM'].str.contains("RAM",na=False)]["RAM"].str.split("GB").str.get(0).apply(lambda x: int(x) * 1000)

In [ ]:
df.loc[(df['RAM'].str.contains("RAM",na=False)) | (df['RAM'].str.contains("GB",na=False)),"RAM"] = df.loc[(df['RAM'].str.contains("RAM",na=False)) | (df['RAM'].str.contains("GB",na=False)),"RAM"].str.split("GB").str.get(0).apply(lambda x: int(x) * 1000)

In [ ]:
df["RAM"].str.split(" ").str.get(1).value_counts()
df[df['RAM'].isna()]

In [ ]:
df["RAM"] = df["RAM"].astype("int")

### Cleaning the 'Chipset' column

In [ ]:
df["Chipset"].str.split(" ").str.get(0).value_counts()

#### 4nm is not chipset, lets remove it

In [ ]:
df.loc[df['Chipset'].str.contains("4nm"),"Chipset"] = df.loc[df['Chipset'].str.contains("4nm"),"Chipset"].str[4:]

In [ ]:
x = df["Chipset"].str.split(" ").str.get(0) + " " + df["Chipset"].str.split(" ").str.get(1)

In [ ]:
df[df['Chipset'].str.contains("Intel")]
df[df['Chipset'].str.contains("TI")]

In [ ]:
df['Chipset'] = df['Chipset'].str.split(" ").str[0:3].str.join(" ")

### Dropping the 'CPU' column

In [ ]:
df.drop(columns="CPU",inplace=True)

### Cleaning the 'Pixel Density' column

In [ ]:
df['Pixel Density'] = df['Pixel Density'].str.split(" ").str.get(0)
df['Pixel Density'] = df['Pixel Density'].astype("int")

### Cleaning the 'Internal Memory' column
* In internal memory, there are values in MB, GB and TB, Lets convert them in one scale.

In [ ]:
df.loc[df['Internal Memory'].str.contains("GB",na=False),"Internal Memory"] = df[df['Internal Memory'].str.contains("GB",na=False)]["Internal Memory"].str.split(" ").str.get(0).apply(lambda x: float(x)*1000)
df.loc[df['Internal Memory'].str.contains("TB",na=False),"Internal Memory"] = df[df['Internal Memory'].str.contains("TB",na=False)]["Internal Memory"].str.split(" ").str.get(0).apply(lambda x: float(x)*1000*1000)
df.loc[df['Internal Memory'].str.contains("MB",na=False),"Internal Memory"] = df[df['Internal Memory'].str.contains("MB",na=False)]["Internal Memory"].str.split(" ").str.get(0)

In [ ]:
df['Internal Memory'] = df['Internal Memory'].astype("int")

### Cleaning the 'Expandable Memory' column

In [ ]:
df['Expandable Memory'] = df['Expandable Memory'].str.split(" ").str.get(0).str.split(",").str.get(0)

### Cleaning the 'SIM Slot' column

In [ ]:
df['SIM Slot'] = df["SIM Slot"].str.split(",").str.get(0)

In [ ]:
df[~(df['SIM Slot'].isin(["Dual SIM","Triple SIM","Single SIM"]))]

* There are just 4 rows with no info about sim slot, we will fill them manually by searching from web 
* It seems all these 4 phones have Single SIM only, as they are Very old Models

In [ ]:
#Filling Rows with empty SIm slot as Single SIM
df.loc[~(df['SIM Slot'].isin(["Dual SIM","Triple SIM","Single SIM"])),'SIM Slot'] = "Single SIM"

In [ ]:
#There is a chronological order in Sim slot, so we can map them in number.
df['SIM Slot'] = df['SIM Slot'].map({"Dual SIM" : 2,"Single SIM" : 1, "Triple SIM" : 3})

In [ ]:
# Converting the column into integer
df['SIM Slot'] =  df['SIM Slot'].astype("int")

### Cleaning the 'WiFi' column

In [ ]:
df["WiFi"] = df["WiFi"].str.split(" ").str.get(0).str.split(",").str.get(0)

### Cleaning the 'Bluetooth' column

In [ ]:
df["Bluetooth"] = df["Bluetooth"].str.split(" ").str.get(0).str.split(",").str.get(0)

### Cleaning the 'GPS' column

In [ ]:
df['GPS'] = df['GPS'].str.split(" ").str.get(0).str.split(",").str.get(0)

### Cleaning the 'Radio' column

In [ ]:
df["Radio"] = df["Radio"].str.split(" ").str.get(0).str.split(",").str.get(0)

In [ ]:
#I dont know what is RDS, will have to check
df[df["Radio"]=="RDS"]

In [ ]:
#RDS is some kind of information about Radio, so it means this phone has Radio
df.loc[df["Radio"]=="RDS","Radio"]="Yes"

### Cleaning the 'Resoulation' column
* We can create 2 cols from Resolution, length and width

In [ ]:
l= df['Resolution'].str.split("x").str.get(0)
df.insert(7,'resolution_length',l)

In [ ]:
w = df['Resolution'].str.split("x").str.get(1).str.split(" p").str.get(0)
df.insert(8,'resolution_width',w)

In [ ]:
df['resolution_width'] = df['resolution_width'].astype("int")
df['resolution_length'] = df['resolution_length'].astype("int")

In [ ]:
#Now we can drop original Resolution Column
df.drop(columns="Resolution",inplace=True)

### Cleaning the 'Screen Size' column
* Same information is given in 2 formats, lets just keep it in Inches

In [ ]:
df['Screen Size'] = df['Screen Size'].str.split(" ").str.get(0)

In [ ]:
df['Screen Size'] = df['Screen Size'].astype("float")

----------------------

## The scrapped data has been cleaned 
#### Converting the cleaned dataset into csv file

In [ ]:
df.to_csv("final_cleaned.csv",index=False)